rust/src/ike/ike.rs

```diff
@@ -54,547 +62,402 @@ impl IkeEvent {
+            let tx = &self.transactions[index];
+            if tx.tx_id < min_tx_id + 1 {
+                index += 1;
+                continue;
+            }
+            *state = index as u64;
+
+            return Some((tx, tx.tx_id - 1, (len - index) > 1));
         }
+
+        return None;
     }
+}
 
-    fn free(&mut self) {
-        if self.events != std::ptr::null_mut() {
-            core::sc_app_layer_decoder_events_free_events(&mut self.events);
-        }
-        if let Some(state) = self.de_state {
-            core::sc_detect_engine_state_free(state);
+/// Probe to see if this input looks like a request or response.
+fn probe(input: &[u8], direction: u8, rdir: *mut u8) -> bool {
+    match parse_isakmp_header(input) {
+        Ok((_, isakmp_header)) => {
+            if isakmp_header.maj_ver == 1 {
+                if isakmp_header.resp_spi == 0 && direction != STREAM_TOSERVER {
+                    unsafe {
+                        *rdir = STREAM_TOSERVER;
+                    }
+                }
+                return true;
+            } else if isakmp_header.maj_ver == 2 {
+                if isakmp_header.min_ver != 0 {
+                    SCLogDebug!(
+                        "ipsec_probe: could be ipsec, but with unsupported/invalid version {}.{}",
+                        isakmp_header.maj_ver,
+                        isakmp_header.min_ver
+                    );
+                    return false;
+                }
+                if isakmp_header.exch_type < 34 || isakmp_header.exch_type > 37 {
+                    SCLogDebug!("ipsec_probe: could be ipsec, but with unsupported/invalid exchange type {}",
+                           isakmp_header.exch_type);
+                    return false;
+                }
+                if isakmp_header.length as usize != input.len() {
+                    SCLogDebug!("ipsec_probe: could be ipsec, but length does not match");
+                    return false;
+                }
+
+                if isakmp_header.resp_spi == 0 && direction != STREAM_TOSERVER {
+                    unsafe {
+                        *rdir = STREAM_TOSERVER;
+                    }
+                }
+                return true;
+            }
+
+            return false;
         }
+        Err(_) => return false,
     }
 }
 
-impl Drop for IKETransaction {
-    fn drop(&mut self) {
-        self.free();
+// C exports.
+export_tx_get_detect_state!(rs_ike_tx_get_detect_state, IKETransaction);
+export_tx_set_detect_state!(rs_ike_tx_set_detect_state, IKETransaction);
+
+/// C entry point for a probing parser.
+#[no_mangle]
+pub extern "C" fn rs_ike_probing_parser(
+    _flow: *const Flow, direction: u8, input: *const u8, input_len: u32, rdir: *mut u8,
+) -> AppProto {
+    if input_len < 28 {
+        // at least the ISAKMP_HEADER must be there, not ALPROTO_UNKNOWN because over UDP
+        return unsafe { ALPROTO_FAILED };
+    }
+
+    if input_len > 1 && input != std::ptr::null_mut() {
```

Comment:
`input_len` already validated to be `>= 28` above.

Reply:
Correct, we can remove the `input_len > 1` part of this check.